### PR TITLE
Move --disable-isolation argument from launch_testing to ros2test

### PR DESCRIPTION
**Pull request by ivanpauno**
Depends on https://github.com/ros2/ament_cmake_ros/pull/3

### CODE REVIEW

File: ros2test/ros2test/command/test.py [hunk 1]
```diff
@@ -12,21 +12,46 @@
...
+        )
 
     def main(self, *, parser, args):
         """Entry point for CLI program."""
+        # If ROS_DOMAIN_ID is already set, launch_test will respect that domain ID and use it.
+        # If ROS_DOMAIN_ID is not set, launch_test will pick a ROS_DOMAIN_ID that's not being used
+        # by another launch_test process.
+        # This is to allow launch_test to run in parallel and not have ROS cross-talk.
+        # If the user needs to debug a test and they don't have ROS_DOMAIN_ID set in their
+        # environment they can disable isolation by passing the --disable-isolation flag.
```
Comment (hidmic) [inline ros2test/ros2test/command/test.py]: @ivanpauno I'd try to condense much of these comments in the associated argument help string above.

File: ros2test/ros2test/command/test.py [hunk 2]
```diff
@@ -12,21 +12,46 @@
...
+        if 'ROS_DOMAIN_ID' not in os.environ and not args.disable_isolation:
+            domain_id = get_coordinated_domain_id()  # Must keep this as a local to keep it alive
+            _logger_.info('Running with ROS_DOMAIN_ID {}'.format(domain_id))
```
Comment (hidmic) [inline ros2test/ros2test/command/test.py]: @ivanpauno consider just printing this given that we are in a CLI command.

**Comment by pbaughman** [comment]
@hidmic Yeah - I have the [ament_cmake PR](https://github.com/ament/ament_cmake/pull/174) open now, and I'll have the `ament_cmake_ros` PR up later today

**Comment by pbaughman** [comment]
@hidmic Follow-up: the ament_cmake_ros PR with the domain_coordinator is here https://github.com/ros2/ament_cmake_ros/pull/3 awaiting review

File: ros2test/ros2test/command/test.py [hunk 3]
```diff
@@ -24,9 +28,25 @@ class TestCommand(CommandExtension):
     def add_arguments(self, parser, cli_name):
         """Add arguments to argparse."""
         launch_testing.launch_test.add_arguments(parser)
+        parser.add_argument(
+            '--disable-isolation', action='store_true', default=False,
+            help='Disable automatic ROS_DOMAIN_ID isolation.'
+            'If ROS_DOMAIN_ID is already set, launch_test will respect and use it.'
+            'If it is not set, a random ROS_DOMAIN_ID that is not being used by another'
```
Comment (pbaughman) [inline ros2test/ros2test/command/test.py]: This is subtly misleading.  The ROS_DOMAIN_ID isn't selected randomly, and it will only coordinate with other processes that use the same cooperative mechanism of coordination.  Maybe a better phrasing is
```
If it is not set, a ROS_DOMAIN_ID not being used by another ros2 test will be picked
```
Probably a nitpick, but I'm worried that the help message makes a promise that the implementation can't deliver on

Comment (ivanpauno) [reply ros2test/ros2test/command/test.py]: Corrected in https://github.com/ros2/ros_testing/pull/2/commits/011c5c2dc628810eb097ca48215782e9f40e1126

**Comment by ivanpauno** [comment]
I tested this, together with https://github.com/ros2/launch/pull/256, https://github.com/ros2/ament_cmake_ros/pull/3 and https://github.com/ament/ament_cmake/pull/174, and the environment variable it's being set correctly.
I don't have an environment for checking if isolation is actually working or not.
@pbaughman Could you test if this is working for you? You will have to change `add_launch_test` with `add_ros_test` for having isolated behavior again:
https://github.com/ros2/ros_testing/blob/44650e46eea70bbd8f879946305965106416f2e4/ros_testing/cmake/add_ros_test.cmake#L61-L81.

File: ros2test/ros2test/command/test.py [hunk 4]
```diff
@@ -24,9 +28,24 @@ class TestCommand(CommandExtension):
     def add_arguments(self, parser, cli_name):
         """Add arguments to argparse."""
         launch_testing.launch_test.add_arguments(parser)
+        parser.add_argument(
+            '--disable-isolation', action='store_true', default=False,
+            help='Disable automatic ROS_DOMAIN_ID isolation.'
+            'If ROS_DOMAIN_ID is already set, launch_test will respect and use it. If it is not'
```
Comment (hidmic) [inline ros2test/ros2test/command/test.py]: @ivanpauno `launch_test` ? :)

File: ros2test/ros2test/command/test.py [hunk 5]
```diff
@@ -24,9 +28,24 @@ class TestCommand(CommandExtension):
...
+            ' set, a ROS_DOMAIN_ID not being used by another ros2 test will be picked. '
```
Comment (hidmic) [inline ros2test/ros2test/command/test.py]: @ivanpauno consider `used` instead of `picked`.

File: ros2test/ros2test/command/test.py [hunk 6]
```diff
@@ -24,9 +28,24 @@ class TestCommand(CommandExtension):
...
+            'When passing this argument, the random ROS_DOMAIN_ID picking is disabled.'
```
Comment (hidmic) [inline ros2test/ros2test/command/test.py]: @ivanpauno what about 
```suggestion
           unless isolation is disabled.'
```
so that it links with the previous sentence.

File: ros2test/ros2test/command/test.py [hunk 7]
```diff
@@ -24,9 +28,24 @@ class TestCommand(CommandExtension):
...
     def main(self, *, parser, args):
         """Entry point for CLI program."""
+        if 'ROS_DOMAIN_ID' not in os.environ and not args.disable_isolation:
+            domain_id = get_coordinated_domain_id()  # Must keep this as a local to keep it alive
+            print('Running with ROS_DOMAIN_ID {}'.format(domain_id))
+            os.environ['ROS_DOMAIN_ID'] = str(domain_id)
+        if 'ROS_DOMAIN_ID' in os.environ:
+            print('ROS_DOMAIN_ID', os.environ['ROS_DOMAIN_ID'])
+        else:
+            print('No ROS_DOMAIN_ID')
```
Comment (hidmic) [inline ros2test/ros2test/command/test.py]: @ivanpauno consider dropping this `print`.

File: ros2test/ros2test/command/test.py (hunk 5 above)
Comment (ivanpauno) [reply ros2test/ros2test/command/test.py]: `...used by another ros2 test will be used.`
I don't like repeating `used`. Maybe `chosen`?

**Comment by ivanpauno** [comment]
CI (only fastrtps):
* Linux [Build Status](https://ci.ros2.org/job/ci_linux/7330/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/3497/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/6007/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/7153/)

**Comment by pbaughman** [comment]
@ivanpauno I manually performed the following test sequence using the ros2 nightly docker image:

1) Build and test demo_nodes_cpp with `master` launch and `master` ros_testing -> Verify that I don't see anything in another terminal when I do ```ros2 topic echo /chatter std_msgs/String``` in one terminal and ```colcon test --packages-select demo_nodes_cpp```
2) Perform a local merge of https://github.com/ros2/launch/pull/256 into launch to remove ROS_DOMAIN_ID isolation -> Verify that I **do** see messages on the /chatter topic because launch_test no longer provides ROS_DOMAIN_ID isolation
3) Perform a  local merge of this PR into ros_testing `master` and update demo_nodes_cpp to use add_ros_test -> Verify that I'm back to not seeing messages on the /chatter topic

Additional cases to check for regressions:
4) Set ROS_DOMAIN_ID to '55' in both of my terminals -> I can see messages on the /chatter topic again
5) Instead of using colcon test, source ws/install/setup.sh and run ```ros2 test /ws/build/demo_nodes_cpp/test_talker_listener__rmw_fastrtps_dynamic_cpp_.py ``` -> I can see messages on the /chatter topic
6) Unset the ROS_DOMAIN_ID in both terminals and run with `ros2 test` -> I can't see messages on the /chatter topic
7) Unset the ROS_DOMAIN_ID and run with ros2 test --disable-isolation -> I can see messages on the /chatter topic

#### In conclusion
Looks good to me
